### ISSUE TITLE
Code signing and other post processing hooks

### ISSUE THREAD

**Issue by bschommer** [ocaml/dune#2861]
I would like to run an action after generating the binary from an executable, for example to perform code signing under windows, therefore I would like to have some action which is performed before installation.

**Comment by nojb** [comment]
One approach may be to **not** make your executable public and instead define an ad-hoc rule that produces the final executable + an explicit `install` stanza.

**Comment by bschommer** [comment]
How would that look like, given for example my current rules is:
````
(executable
 (names foo)
 (public_names bar)
...)
````

**Comment by nojb** [comment]
You would remove `(public_name bar)` and add two extra stanzas:
```
(rule
 (target bar.exe)
 (action (run <sign> -o %{target} %{dep:foo.exe}))

(install
 (section bin)
 (package <your-package>)
 (files (./bar.exe as bar)))
```
Note this is off the top my head -- it needs to be tested :)

**Comment by bschommer** [comment]
That would work, however that is exactly what I tried to avoid.

**Comment by bschommer** [comment]
What would really help to avoid such boilerplate code, also for other stuff would be a way to write some kind of macro that expands to several such stanzas. This would also help with some other problems we currently face, for example more complex tests that require more than one action per test case.

**Comment by diml** [comment]
Definitely agreed on that point. Better support for user defined stanzas is something we have been talking about for a while, but haven't really prioritised.

**Comment by bschommer** [comment]
I think having such macros or functions that expand to stanzas would be really nice and would allow it to easier write extensions etc and I would be willing to help implement such a feature, but I guess one first needs to have some kind of idea what features should be supported and what not.

**Comment by diml** [comment]
Help would be nice :)

I'd say the main feature we want to provide is better support for factorising and sharing the contents of `dune` files. One idea would be to extend the `dune` language to support functions, etc... but I feel like we'd just end up designing a bad programming language. Another approach would be to allow defining custom stanzas via OCaml code.

**Comment by rgrinberg** [comment]
@diml didn't we have an approach on how to define user stanzas in OCaml? I recall that you said it wasn't the way to go but it seemed like a promising approach to me.

I'm also against implementing a programming language in dune.

**Comment by diml** [comment]
Yeah, I wrote a proposal about preprocessing dune files. But it feels a bit too hacky. I guess we could adapt it to still allow defining user stanzas in OCaml. It's a bit difficult to find the right design, but I guess we should start somewhere.

Here is a question floating in my head:  should extensions, i.e. user defined stanzas implemented in OCaml that produces vanilla stanzas, have access to build artifacts? For instance, should an extension be allowed to use Base or an image parsing library, ...?

There are cases where it could be useful, for instance you could imagine parsing a file in a complex format using some library and produce rules based on that. At the same time it does complicate the implementation. It's also more opportunities for extensions to break for one reason or another.

Another question is the following: how should Dune communicate with extensions? The easiest model would be to have the extensions statically linked in Dune. Then you could effectively use whatever you want in the extensions. However, it would mean that you couldn't stat working on a project using extensions by just typing `dune`. You'd first have to build a custom dune and then use it.

Relying on dynlink is not really an option if we want `dune` to not be tied to a particular version of OCaml. Plus, there are always portability problems around dynlinking things.

If you accept that the code of extensions only have access to a restricted set of OCaml APIs, then you can imagine another model where `dune` starts by building the extensions as a full blown executable and then use it to interpret the workspace.

Maybe both approaches, the "custom dune" and the "on the fly" restricted extensions are useful in different cases. But I'm wondering if we need two mechanisms here.

Anyway, these are just my current thoughts on the matter.

**Comment by bschommer** [comment]
I personally don't think that a "custom dune" is the right way to go, since then distributing a project that relies on this such a "custom dune" for example via opam would require either a dependency to a new dune package that contains the extension or you would need to change dune to work like ocamlbuild does and build a dune binary before invoking the actual commands.

**Comment by nojb** [comment]
Wasn't the plan to allow dynamically generating dune files ?

**Comment by diml** [comment]
@bschommer indeed, that's a limitation.

@nojb we'll certainly add this. However, I'm wondering if that's enough for the case considered here.

More precisely, generating dune files will be nice to setup a complex custom testsuites. However, if we want to have custom stanzas that could be mixed with regular ones and that user could share, it might be a bit more complicated.

**Comment by nojb** [comment]
> More precisely, generating dune files will be nice to setup a complex custom testsuites. However, if we want to have custom stanzas that could be mixed with regular ones and that user could share, it might be a bit more complicated.

Not sure this makes total sense, but perhaps one could interpret "user-defined" stanzas as arguments to a program that generates dynamic dune files? The user-defined stanza would identify the particular executable that implements it somehow and dune would call the executable with it as argument to desugar it.

**Comment by diml** [comment]
That makes sense yeah. I'm just worried about performances here as calling a program can be expansive.

**Comment by bschommer** [comment]
> Not sure this makes total sense, but perhaps one could interpret "user-defined" stanzas as arguments to a program that generates dynamic dune files? The user-defined stanza would identify the particular executable that implements it somehow and dune would call the executable with it as argument to desugar it.

That sounds like a reasonable idea, if the "user-defined" stanzas would be added in a separate file, for example `dune.meta` or so and then dune would generate a `dune` file using for example an executable defined in the `dune.meta` and only would need to update the `dune` file itself if for example the `dune.meta` file changed.

**Comment by diml** [comment]
Would you then commit the generated `dune` file? You can already do that with promotion, however the workflow is not super nice.

**Comment by bschommer** [comment]
I'm not 100% sure of the way to go there, however I would opt to not commit the generated `dune` file.

**Comment by rgrinberg** [comment]
There was an offline discussion about plugins that I'll summarize:

## Why dune doesn't have plugins yet?

One of the main reasons why we've held off plugins for so long is that we want to be able to evaluate dune files as early as possible. One reason for this is that we'd prefer to avoid consulting the library database to do evaluate dune files, since creating the library db itself requires us to evaluate dune files. While this problem can be solved, you can see how it would really complicate dune.

## Our compromise

We'll allow user defined plugins, but we'll restrict them from being modular with these restrictions:

* plugins will not be allowed to have library dependencies.
* plugins will have to live a single OCaml (no reasonml) source file.
* plugins can only read source files & directories. Doing these operations records a dependency on these sources.

The net effect of these restrictions is that we'll no longer need to setup any rules to interpret dune files.

## Proposed API

The API will be quite minimal and will only cover 2 aspects: reading source files and defining stanzas. I've omitted all versioning for brevity reasons. Here's a draft of the file system API:

```ocaml
module Source : sig
  (** Only reading source files is allowed. *)

  module Path : sig
    type t
    val to_string : t -> string
    val parent : t -> t option
    val relative : t -> string -> t
  end

  module File : sig
    type t
    val path : Path.t
    val read : t -> string
  end

  module Dir : sig
    type t
    val path : t -> Path.t
    val files : t -> glob:Glob.t -> File.t list
    val subdirs : t -> glob:Glob.t -> Dir.t list
  end

  val workspace_root : unit -> Path.t

  val project_root : unit -> Path.t

  val find_dir : Path.t -> Dir.t option
end
```

Note that listing directories and reading files will record a dependency on those files.

To parse & emit stanzas, we need a simple API. What I have below is essentially a slimmed down of `Dune_lang`:

```ocaml

module Dune_lang : sig
  type t = Atom of Loc.t * string | List of Loc.t * string ..

  module Loc : sig
    type t
  end

  module Decoder : sig
    type 'a t
    type dune_lang

    (* minimal monadic api for decoding stanzas *)

    val decode : 'a t -> dune_lang -> 'a
  end with type dune_lang := t

  module Encoder : sig
    type t
    val string : string -> t
    val record : (string * t) -> t
    val list : t list -> t
  end
end
```

For the user to register and define the behavior of their plugin they these 2 modules:

```
module Stanza : sig
  type t
  val loc : t -> Loc.t
  val name : t -> string
  val contents : t -> Dune_lang.t
end

module Context : sig
  type t
  val version : t -> int * int
  val stanza : t -> Stanza.t
  val error : Loc.t -> string -> 'a
  val dir : t -> Source.Dir.t
  val emit : t -> Dune_lang.Encoder.t -> unit

  (* This is essentially a rule to generate a source file in the current 
     directory. This is quite useful for setting up source code boilerplate *)
  val write_file : t -> name:string -> contents:string -> unit
end

val register : (Context.t -> unit) -> unit
```

Suggestions to the API above are welcome. But let's strive for something that is as simple as possible to implement. I'm convinced that even something primitive is going to cover the majority of use cases. Once we gather some user experiences, we can always publish a V2 of the interface.

## Using plugins

Plugins must can be declared in the dune-project file. I've separated declaration from usage, because I expect plugins to be usable cross-project one day. Also, it makes it easier for a plugin to define multiple versions of itself.

```
(plugin
 (name plug)
 (source bar/foo)) ;; corresponds to bar/foo.ml

;; to enable the plugin, we need this line.
(using plug 0.1)
```

Now all stanzas of the form `plug.foo` will be passed to `foo.ml` for interpretation.

**Comment by bschommer** [comment]
I think it would be better if the plugin would have a different file suffix in order to avoid clashes with `*.ml` files. I additionally would allow the plugin directory to be specified by a command line option, this would easily allow reusing and distributing such plugins between multiple projects, maybe even add a default location were dune searches for such plugins.

**Comment by rgrinberg** [comment]
@aalekseyev could you give some feedback on the API?

**Comment by rgrinberg** [comment]
> I think it would be better if the plugin would have a different file suffix in order to avoid clashes with *.ml files.

How about a `dune.ml` suffix? This will get picked up automatically by editors.

> I additionally would allow the plugin directory to be specified by a command line option, this would easily allow reusing and distributing such plugins between multiple projects, maybe even add a default location were dune searches for such plugins.

That's a natural extension, but I think it would be best to mimc the model for dune libraries. This would mean that one could specify a `public_name` for a plugin, and this would make it usable in other projects. Otherwise, we could simply allow qualified names of the form `project.plugin_name` in the `using` stanza. We'd still need a way to mark private/public plugins though.

**Comment by aalekseyev** [comment]
One thought is that if plugins are allowed to produce arbitrary stanzas, then in a large project you need to execute *all* plugins before building anything, so that you can resolve library dependencies.

Another thought is that the limitation that you can't use libraries is a really brutal limitation. I think we should try hard to relax that. Here's one brainstorm-quality idea: we can do library resolution per-project, and a plugin is allowed to use public libraries from other projects.

**Comment by rgrinberg** [comment]
> One thought is that if plugins are allowed to produce arbitrary stanzas, then in a large project you need to execute all plugins before building anything, so that you can resolve library dependencies.

Note that when we execute plugins, we record the sources that they read. This will be enough to "cutoff" their execution in many cases. Hopefully this is enough to make the performance penalty less severe.

> Another thought is that the limitation that you can't use libraries is a really brutal limitation.

Yeah, this is indeed quite limiting. I expect that we'll lift this in the long term.

> Here's one brainstorm-quality idea: we can do library resolution per-project, and a plugin is allowed to use public libraries from other projects.

The problem with this is that we now have to interleave populating the library database and evaluating the dune stanzas. That complicates the implementation quite a bit. Although I certainly think that it's worth doing eventually.

**Comment by bschommer** [comment]
>How about a .dune.ml suffix? This will give syntax highlighting automatically in all editors.
Sounds good, however I then would exclude this from being included in libraries/executables in the same directory (if not explicitly specified) in so that these files can life in the same directories or sub-directories if included.

>EDIT: @bschommer if possible, let's consider sharing plugins between different projects as out of scope for now.
Fine with me.

**Comment by bschommer** [comment]
Should I change the title so that others find this issue easier?

**Comment by rgrinberg** [comment]
> Sounds good, however I then would exclude this from being included in libraries/executables in the same directory (if not explicitly specified) in so that these files can life in the same directories or sub-directories if included.

Since these aren't valid module names, dune will not pick them up for executables & libraries.

**Comment by diml** [comment]
Personally, I'd keep the .ml extension and just put the plugin in a separate directory.

Just a couple of thoughts: the proposal seems good to me, and I agree that eventually we should allow to use libraries in plugins. Doing so will indeed require quite a bit of work on Dune's internals. Since users have wanted this feature for a long time and it would unlock whole new world of possibilities, it's nice if we can provide a first answer now.

The key is to keep our options open so that we are not stuck in the future. It seems to me that this is the case here. For instance we could allow plugins to use libraries using @aalekseyev's idea. There is however one case where it might be difficult: when a plugin generates a stanzas that installs a public binary. This is because public binary names don't reflect what packages they are from. So for now I propose to forbid plugins from generating stanzas that install public binaries.

I think we can also allow plugins to use each other without limiting our options. This would be an easy way to at least share code between plugins.

Finally, we also discussed quickly the possibility of having dune interpret the language itself rather than compile and execute. I don't know if this is viable, but this would have several advantages: this would keep `dune` properly self contained and would allow dune to detect when a function is modified, which could be helpful to memoise things in a more fine grained manner. The current proposal would be mostly compatible with this idea, except for the fact that we'd have to lift the whole OCaml standard library into the interpreted world, which doesn't seem realistic. So if we want to keep this option open, we have to hide the standard library entirely from plugins and instead provide our own small stdlib to plugins. Thoughts?

**Comment by rgrinberg** [comment]
>  So for now I propose to forbid plugins from generating stanzas that install public binaries.
Could you clarify this limitation a little more? Is the issue that we need to evaluate all plugins before resolving binaries?

> I think we can also allow plugins to use each other without limiting our options. This would be an easy way to at least share code between plugins.
That’s fine, but I think this falls outside the MVP. We can this a bit later.
> So if we want to keep this option open, we have to hide the standard library entirely from plugins and instead provide our own small stdlib to plugins. Thoughts?
I think that if we version our plugin API then we keep our options even more open without adding this limitation. We can always say that V2 of the plugin API is not going to have Stdlib built in. I think it would be quite optimistic if we could upgrade everyone’s plugins to our own interpreted OCaml without any kind of breakage anyway, so it makes sense to me if both plugin api’s could live in parallel at least in the medium term.

**Comment by diml** [comment]
> Could you clarify this limitation a little more? Is the issue that we need to evaluate all plugins before resolving binaries?

Yes, that the problem.

> That’s fine, but I think this falls outside the MVP. We can this a bit later.

Agreed.

> I think that if we version our plugin API then we keep our options even more open without adding this limitation. We can always say that V2 of the plugin API is not going to have Stdlib built in. I think it would be quite optimistic if we could upgrade everyone’s plugins to our own interpreted OCaml without any kind of breakage anyway, so it makes sense to me if both plugin api’s could live in parallel at least in the medium term.

I think that's ok if we mark the plugin system as exprimental. Otherwise, we should really have the samllest possible API. That will give us more flexibility moving forward. And the stdlib is a large API.

**Comment by rgrinberg** [comment]
As discussed offline, we've decided on the following:

* The dune plugin will in fact be available as a normal library from the `dune-plugin` package. Plugins will compile and link against this (and only this) library.

* The stdlib will not be available to users. We'll provide a minimal standard library for users.

The reasoning for these changes is that we'd like to leave the option to use an embedded OCaml to run plugins.

A question for @diml regarding plugins:

If plugins are all compiled and linked together, then we'll need a library database just for plugins. This means that we'll need a special "plugin" scope where only plugins and plugin API are available.
